src/OpenLoco/Vehicles/Routing.cpp

```diff
@@ -0,0 +1,670 @@
+                    }
+                }
+                if (animate)
+                {
+                    Map::AnimationManager::createAnimation(0, signalLoc, signalLoc.z / 4);
+                    Ui::ViewportManager::invalidate(signalLoc, signalLoc.z, signalLoc.z + 24, ZoomLevel::half);
+                }
+            }
+        }
+        return loc.x; // Odd???
+    }
+
+    // 0x004A2AF0
+    // Passes occupied state via 0x001135F88
+    // Returns true for signals
+    static bool findSignalsAndOccupation(const LocationOfInterest& interest)
+    {
+        auto nextLoc = interest.loc;
+        const auto tad = interest.tad();
+        auto& trackSize = Map::TrackData::getUnkTrack(tad._data);
+        nextLoc += trackSize.pos;
+        if (trackSize.rotationEnd < 12)
+        {
+            nextLoc -= Map::Pos3{ _503C6C[trackSize.rotationEnd], 0 };
+        }
+
+        auto backwardTaD = tad;
+        backwardTaD.setReversed(!backwardTaD.isReversed());
+        const auto startLoc = tad.isReversed() ? nextLoc : interest.loc;
+
+        for (const auto& trackPiece : TrackData::getTrackPiece(tad.id()))
+        {
+            const auto rotPos = Math::Vector::rotate(Pos2{ trackPiece.x, trackPiece.y }, tad.cardinalDirection());
+            const auto trackLoc = Pos2{ startLoc } + rotPos;
+
+            for (auto* entity : EntityManager::EntityTileList(trackLoc))
+            {
+                auto* vehicle = entity->asVehicle();
+                if (vehicle == nullptr)
+                {
+                    continue;
+                }
+
+                if (vehicle->getFlags38() & (Vehicles::Flags38::unk_0 | Vehicles::Flags38::unk_2))
+                {
+                    continue;
+                }
+
+                if (vehicle->getTrackLoc() == interest.loc && vehicle->getVar2C().track == tad)
+                {
+                    addr<0x001135F88, uint16_t>() = 1;
+                    break;
+                }
+
+                if (vehicle->getTrackLoc() == nextLoc && vehicle->getVar2C().track == backwardTaD)
+                {
+                    addr<0x001135F88, uint16_t>() = 1;
+                    break;
+                }
+            }
+        }
+        return interest.trackAndDirection & (1 << 15);
+    }
+
+    // 0x004A2CE7
+    // Passes occupied state via 0x001135F88
+    static void setSignalsOccupiedState(const LocationOfInterestHashMap& hashMap)
+    {
+        for (const auto& interest : hashMap)
+        {
+            if (!(interest.trackAndDirection & (1 << 15)))
+            {
+                continue;
+            }
+
+            bool isOccupied = addr<0x001135F88, uint16_t>() & 1;
+            uint32_t flags = (1ULL << 31) | (isOccupied ? 8ULL : 9ULL);
+            sub_48963F(interest.loc, interest.tad(), interest.trackType, flags);
+        }
+    }
+
+    // 0x004A2D4C
+    static bool sub_4A2D4C(const LocationOfInterest& interest)
+    {
+        if (!(interest.trackAndDirection & (1 << 15)))
```

Comment:
Lots of cases of this. I am a bit confused because it looks like we have a dedicated struct for TrackAndDirection with helper methods that could be used instead, so shouldn't that be used everywhere?

Reply:
This is looking for a signal element it's outside of the normal track and direction

Reply:
This is looking for a signal element it's outside of the normal track and direction

Reply:
Still a bit confused, do we need two different structs for this then?

Reply:
I was sort of thinking of 3 classes. One for the basic TaD one for TaDandSignal and one for everything else. That appears to be how it is used